experimental/InjectiveResolutions/src/MonoidAlgebra.jl

```diff
@@ -0,0 +1,611 @@
+  return MonoidAlgebraIdeal(base_ring(I), radical(underlying_ideal(I)))
+end
+
+dim(I::MonoidAlgebraIdeal) = dim(underlying_ideal(I))
+
+# some generic functionality which should probably be elsewhere
+function is_subset(I::T, J::T) where {T<:Ideal}
+  return all(x in J for x in gens(I))
+end
+
+function Base.:(==)(I::T, J::T) where {T <: Ideal}
+  return is_subset(I, J) && is_subset(J, I)
+end
+
+function Base.:*(I::T, J::T) where {T<:Ideal}
+  @assert base_ring(I) === base_ring(J)
+  return ideal(base_ring(I), [x*y for x in gens(I) for y in gens(J)])
+end
+
+# user facing constructor
+ideal(A::MonoidAlgebra, v::Vector) = MonoidAlgebraIdeal(A, elem_type(A)[A(x) for x in v])
+
+function Base.in(a::MonoidAlgebraElem, I::MonoidAlgebraIdeal)
+  return underlying_element(a) in underlying_ideal(I)
+end
+
+function coordinates(a::MonoidAlgebraElem, I::MonoidAlgebraIdeal)
+  return coordinates(underlying_element(a), underlying_ideal(I))
+end
+
+function intersect(a::MonoidAlgebraIdeal, b::MonoidAlgebraIdeal...)
+  kQ = base_ring(a)
+  @req all(base_ring(g) === kQ for g in b) "base rings must match"
+  as = Singular.intersection(singular_generators(a), [singular_generators(g) for g in b]...)
+  return MonoidAlgebraIdeal(kQ, [kQ(x) for x in gens(as)])
+end
+
+function Base.intersect(V::Vector{T}) where {T <: MonoidAlgebraIdeal}
+  @assert length(V) != 0
+  length(V) == 1 && return V[1]
+
+  return intersect(V[1], V[2:end]...)
+end
+
+# random elements for testing 
+rand(A::MonoidAlgebra, v...) = A(rand(A.algebra, v...))
+
+### Additional functionality for ring conformance tests
+Base.hash(a::MonoidAlgebraElem, h::UInt) = hash(underlying_element, h)
+characteristic(A::MonoidAlgebra) = characteristic(A.algebra)
+divexact(a::MonoidAlgebraElem, b::MonoidAlgebraElem; check::Bool=true) = parent(a)(divexact(underlying_element(a), underlying_element(b); check))
+
+function divides(a::MonoidAlgebraElem, b::MonoidAlgebraElem)
+  success, q = divides(underlying_element(a), underlying_element(b))
+  return success, parent(a)(q)
+end
+
+is_nilpotent(a::MonoidAlgebraElem) = is_nilpotent(underlying_element(a))
+canonical_unit(a::MonoidAlgebraElem) = canonical_unit(underlying_element(a))
+is_domain_type(::Type{MonoidAlgebraElem{CT, PT}}) where {CT, AT, PT <: MonoidAlgebra{CT, AT}} = is_domain_type(elem_type(AT))
+# dummy method required by the conformance test
+#divrem(a::MonoidAlgebraElem, b::MonoidAlgebraElem; check::Bool=true) = zero(a), a
+
+function monoid_algebra_ideal(kQ::MonoidAlgebra, I::Ideal)
+  @req base_ring(I) == kQ.algebra "base rings do not match"
+  return MonoidAlgebraIdeal(kQ, I)
+end
+
+function Oscar.quotient_ring_as_module(I::MonoidAlgebraIdeal)
+  R = base_ring(I)
+  F = graded_free_module(R,1)
+  e1 = F[1]
+  return quo_object(F, [x * e1 for x = gens(I)]) 
+end
+
+function Base.show(io::IO, ::MIME"text/plain", I::MonoidAlgebraIdeal)
+  println(
+    io, "ideal over $(base_ring(I)) generated by "*join(["$(x)" for x in gens(I)], ", ")
+  )
```

Comment:
`join` can target an `io` object which is more efficient then first printing into a string. So this should do it:
```suggestion
  print(pretty(io), "Ideal over ", Lowercase(), base_ring(I), " generated by "
  join(io, gens(I), ", ")
```